compiler/rustc_passes/src/dead.rs

```diff
@@ -814,6 +814,13 @@ impl<'tcx> DeadVisitor<'tcx> {
             }
         };
 
+        for id in &dead_codes[1..] {
+            let hir = self.tcx.hir().local_def_id_to_hir_id(*id);
+            let lint_level = self.tcx.lint_level_at_node(lint::builtin::DEAD_CODE, hir).0;
+            if let Some(expectation_id) = lint_level.get_expectation_id() {
+                self.tcx.sess.diagnostic().insert_fulfilled_expectation(expectation_id);
+            }
+        }
```

Comment:
I suggest to refactor the code the other way around.
We should assert that all ids in `dead_codes` have the same lint level.
And callers should group calls by level, like `warn_dead_fields_and_variants`.
The branch of `check_mod_deathness` for impl items may use `warn_dead_fields_and_variants` instead of `warn_multiple_dead_codes` directly.